### PR TITLE
Send register via AMQP

### LINKED ISSUE #10: Send 'register' message to message queue
- [ ] Replace msg_register to use rabbitmq
- [ ] Listen to exchange devices.registered
- [ ] Send ID inside UUID filed

### PR DESCRIPTION
**Pull request by ramonhpr**
This PR closes #10 

How to test:
install pika python package
run: test/mock-connector.py
run knotd: sudo src/knotd -nr -c src/knotd.conf
register a thing with thing mocked script: test/test-conn.py --debug

Expected result:
the mock-connector should show the messages in json
the test-conn.py should receive unregister request and show
And knotd should add the dbus device in d-feet

### CODE REVIEW

File: src/cloud.c [hunk 1]
```diff
@@ -65,6 +66,7 @@ union cloud_cb_t {
 	cloud_downstream_cb_t update_cb;
 	cloud_downstream_cb_t request_cb;
 	cloud_device_removed_cb_t removed_cb;
+	cloud_device_added_cb_t added_cb;
```
Comment (vitbaq) [inline src/cloud.c]: For me is more intuitive if the added come before removed.

File: src/msg.c [hunk 2]
```diff
@@ -932,15 +934,13 @@ static ssize_t msg_process(struct session *session,
...
 		if (result != 0)
 			break;
-
-		/* Enable downstream after registration & authentication */
 		session->downstream_to =
 			l_timeout_create_ms(512,
 					    downstream_callback,
 					    session, NULL);
+		return 0;
 		break;
```
Comment (vitbaq) [inline src/msg.c]: Remove break; or return 0;

File: src/msg.c [hunk 3]
```diff
@@ -1299,10 +1318,27 @@ static void on_device_added(const char *device_id, const char *token,
 			return;
 	}
 
+	device_set_uuid(device, mydevice->id);
```
Comment (vitbaq) [inline src/msg.c]: Set device uuid with id? Explain that. You can put a comment here.

File: src/msg.c [hunk 4]
```diff
@@ -1290,34 +1279,76 @@ static void forget_if_unknown(struct knot_device *device, void *user_data)
...
+		hal_log_dbg("Session not found");
+		/*
+		 * TODO: send a nack to amqp so the same message can be handled
+		 * later
+		 */
+		return;
+	}
+
+	node_ops = session->node_ops;
+
```
Comment (vitbaq) [inline src/msg.c]: Remove extra line

File: src/msg.c [hunk 5]
```diff
@@ -1487,7 +1517,7 @@ int msg_start(struct settings *settings)
 		goto cloud_fail;
 	}
 
-	err = cloud_set_read_handlers(on_update_cb, on_request_cb,
+	err = cloud_set_read_handlers(on_update_cb, on_request_cb, on_device_added,
```
Comment (vitbaq) [inline src/msg.c]: More than 80 columns, fix it.

File: src/cloud.h [hunk 6]
```diff
@@ -36,3 +41,4 @@ int cloud_publish_data(const char *id, uint8_t sensor_id,
 		       const knot_value_type *value,
 		       uint8_t kval_len);
 int cloud_unregister_device(const char *id);
+int cloud_register_device(const char *id, const char *name);
```
Comment (vitbaq) [inline src/cloud.h]: Put before cloud_unregister_device, to keep the order followed in cloud.c

File: src/msg.c [hunk 7]
```diff
@@ -551,12 +551,14 @@ static int8_t msg_register(struct session *session,
 		return result;
 
 	mydevice->id = l_strdup(id);
+	mydevice->uuid = l_strdup(id);
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/msg.c]: Ok

File: src/msg.c [hunk 8]
```diff
@@ -1290,34 +1278,31 @@ static void forget_if_unknown(struct knot_device *device, void *user_data)
...
 
-static void proxy_added(const char *device_id, const char *uuid,
-			const char *name, bool online,
-			void *user_data)
+static void on_device_added(const char *device_id, const char *token,
+			    void *user_data)
 {
 	struct knot_device *device = device_get(device_id);
-	struct mydevice *mydevice = l_new(struct mydevice, 1);
+	struct mydevice *mydevice = l_queue_find(device_id_list,
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/msg.c]: device -> device_dbus
mydevice -> device_pending

Something like that

File: src/msg.c (hunk 5 above)
Comment (ramonhpr) [reply src/msg.c]: Done

File: src/cloud.h (hunk 6 above)
Comment (ramonhpr) [reply src/cloud.h]: Done

File: src/msg.c (hunk 4 above)
Comment (ramonhpr) [reply src/msg.c]: Done

File: src/msg.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (ramonhpr) [reply src/msg.c]: Done

File: src/proto.c [hunk 9]
```diff
@@ -287,53 +287,6 @@ void proto_close(int prot_sock)
...
-		 const char *device_name, const char *device_id,
-		 char *uuid, char *token)
-{
-	json_object *device;
-	const char *device_as_string;
-	json_raw_t response;
-	int err, result;
-
-	memset(&response, 0, sizeof(response));
-	device = parser_device_json_create(owner_uuid, device_name, device_id);
```
Comment (alla3) [inline src/proto.c]: This function is not been used anymore. Maybe you should call it in another place or remove it.

File: src/proto.c [hunk 10]
```diff
@@ -287,53 +287,6 @@ void proto_close(int prot_sock)
...
-	json_object_put(device);
-
-	if (err < 0) {
-		hal_log_error("manager mknode: %s(%d)", strerror(-err), -err);
-		result = KNOT_ERR_CLOUD_FAILURE;
-		goto fail;
-	}
-
-	if (response.size == 0 ||
-	    (parser_device(response.data, uuid, token) < 0)) {
```
Comment (alla3) [inline src/proto.c]: This function is not been used anymore. Maybe you should call it in another place or remove it.

Comment (ramonhpr) [reply src/proto.c]: Done

File: src/msg.c [hunk 11]
```diff
@@ -493,12 +495,11 @@ static int8_t msg_register(struct session *session,
 			   const knot_msg_register *kreq, size_t ilen,
 			   knot_msg_credential *krsp)
 {
-	struct knot_device *device;
+	struct mydevice *mydevice = l_new(struct mydevice, 1);
```
Comment (vitbaq) [inline src/msg.c]: What do you think about change mydevice to device_pending?

Comment (ramonhpr) [reply src/msg.c]: Done

**Pull request merged**